src/treelearner/feature_histogram.hpp

```diff
@@ -85,22 +85,25 @@ class FeatureHistogram {
   void FindBestThreshold(double sum_gradient, double sum_hessian,
                          data_size_t num_data,
                          const ConstraintEntry& constraints,
+                         double parent_output,
                          SplitInfo* output) {
     output->default_left = true;
     output->gain = kMinScore;
     find_best_threshold_fun_(sum_gradient, sum_hessian + 2 * kEpsilon, num_data,
-                             constraints, output);
+                             constraints, parent_output, output);
     output->gain *= meta_->penalty;
   }
 
-  template <bool USE_RAND, bool USE_L1, bool USE_MAX_OUTPUT>
-  double BeforeNumercal(double sum_gradient, double sum_hessian,
+  template <bool USE_RAND, bool USE_L1, bool USE_MAX_OUTPUT, bool USE_SMOOTHING>
+  double BeforeNumercal(double sum_gradient, double sum_hessian, double parent_output, data_size_t num_data,
                         SplitInfo* output, int* rand_threshold) {
     is_splittable_ = false;
     output->monotone_type = meta_->monotone_type;
-    double gain_shift = GetLeafGain<USE_L1, USE_MAX_OUTPUT>(
-        sum_gradient, sum_hessian, meta_->config->lambda_l1,
-        meta_->config->lambda_l2, meta_->config->max_delta_step);
+    double output_without_split = CalculateSplittedLeafOutput<USE_L1, USE_MAX_OUTPUT, USE_SMOOTHING>(
+        sum_gradient, sum_hessian, meta_->config->lambda_l1, meta_->config->lambda_l2,
+        meta_->config->max_delta_step, meta_->config->path_smooth, num_data, parent_output);
+    double gain_shift = GetLeafGainGivenOutput<USE_L1>(
```

Comment:
may be a new function that `CalculateSplittedLeafOutput` then `GetLeafGainGivenOutput` ? it seems some other places also use this.

Reply:
Actually I realised we can just use the existing function `GetLeafGain` here.